S 57908 simplified

dillius opened the pull request.

Commit:
S-57908: First pass of refactoring Localization to allow for lists of individual template endpoints, so that file/database sources can be mixed.

Commit:
S-57908: Cleanup and addition of FileTemplateProvider and tests.

pmamut commented on source/LocalizationManager.cs:
```diff
@@ -87,21 +95,19 @@ private Localizer CreateDefaultLocalizer ()
 
 
-		private static Localizer CreateLocalizer (Localizer fallback, string culture, ITemplateSetLoader loader, params string[] setNames)
+		private static Localizer CreateLocalizer(Localizer fallback, string culture, IEnumerable<ITemplateProvider> providers)
```
`params ITemplateProvider[]` FTW!

pmamut commented on source/LocalizationManager.cs:
```diff
@@ -2,28 +2,36 @@
+			params string[] setnames)
+			: this(
+				new CultureInfo(defaultculture),
+				setnames.Select(setname => new CompatibilityTemplateProvider(loader, setname)))
```
`.ToArray()` to materialize the collection once.

pmamut commented on source/LocalizationManager.cs:
```diff
@@ -113,16 +119,17 @@ private static Localizer CreateLocalizer (Localizer fallback, string culture, IT
 				catch (Exception e)
 				{
-					throw new TemplateSetLoadException(culture, setname, e);
+					throw new TemplateSetLoadException(culture, provider.Name, e);
```
Entire `try {} catch () {}` needs to go into `CompatibilityTemplateProvider.Load()`, since that's where culture-specific template set loading responsibility is moved.

This also removes the need for `provider.Name`.

pmamut commented on source/TemplateProvider.cs:
```diff
@@ -0,0 +1,56 @@
+		string Name { get; }
+	}
+
+	public class CompatibilityTemplateProvider : ITemplateProvider
```
Maybe call him `DefaultTemplateProvider` or some such...

pmamut commented on source/TemplateProvider.cs:
```diff
@@ -0,0 +1,56 @@
+		public string Name { get; private set; }
+	}
+
+	public class FileTemplateProvider : ITemplateProvider
```
`FileTemplateProvider` is superfluous, since it just duplicates the code from existing `FileTemplateSetLoader`, and its functionality is simply a binding of `FileTemplateSetLoader` to a set name via `CompatibilityTemplateProvider` instance.

Commit:
Removal of a default implementation for a FileTemplateProvider and its test

Commit:
S-57908: Cleanup refactorings
Exception checking done in ITemplateProvider
ITemplateProviders now provided as params list instead of IEnumerable
Removed Name element from ITemplateProvider

Commit:
S-57908: Fixed Project File and corrected type conversion warning

pmamut commented on source/LocalizationManager.cs:
```diff
@@ -2,28 +2,36 @@
-			: this(new CultureInfo(defaultculture), loader, setnames)
+		public LocalizationManager(string defaultculture, ITemplateSetLoader loader, IDictionary<string, string> overrides,
+			params string[] setnames)
+			: this(
```
Could this simply delegate to the constructor below to DRY up provider set creation?

`: this(new CultureInfo(defaultculture), loader, setnames)`

Commit:
S-57908: Removed repeat conversion of ITemplateSetLoader to ITemplateProviders

Commit:
S-57908: Culture parameter consistency
New localization manager signatures are now consistent with old signatures in handling of Culture.
